src/sql/parts/notebook/models/notebookModel.ts

```diff
@@ -711,45 +785,36 @@ export class NotebookModel extends Disposable implements INotebookModel {
 	}
 
 	/**
-	 * Set _providerId and _activeClientSession based on a kernelSpec representing new kernel
-	 * @param kernelSpec KernelSpec for new kernel
+	 * Set _providerId and start session if it is new provider
+	 * @param displayName Kernel dispay name
 	 */
-	private setProviderIdForKernel(kernelSpec: nb.IKernelSpec): void {
-		if (!kernelSpec) {
-			// Just use the 1st non-default provider, we don't have a better heuristic
-			let notebookManagers = this._notebookOptions.notebookManagers.filter(manager => manager.providerId !== DEFAULT_NOTEBOOK_PROVIDER);
-			if (!notebookManagers.length) {
-				notebookManagers = this._notebookOptions.notebookManagers;
-			}
-			if (notebookManagers.length > 0) {
-				this._providerId = notebookManagers[0].providerId;
-			}
-		} else {
-			let sessionManagerFound: boolean = false;
-			for (let i = 0; i < this.notebookManagers.length; i++) {
-				if (this.notebookManagers[i].sessionManager && this.notebookManagers[i].sessionManager.specs && this.notebookManagers[i].sessionManager.specs.kernels) {
-					let index = this.notebookManagers[i].sessionManager.specs.kernels.findIndex(kernel => kernel.name === kernelSpec.name);
-					if (index >= 0 && this._clientSessions && this._clientSessions.length > 0) {
-						if (this._activeClientSession) {
-							this._oldClientSession = this._activeClientSession;
-						}
-						this._activeClientSession = this._clientSessions[i];
-						if (this.notebookManagers[i].providerId !== this._providerId) {
-							this._providerId = this.notebookManagers[i].providerId;
-							this._onProviderIdChanged.fire(this._providerId);
+	private async setProviderIdAndStartSession(displayName: string): Promise<void> {
+		if (displayName) {
+			if (this._activeClientSession && this._activeClientSession.isReady) {
+				this._oldKernel = this._activeClientSession.kernel;
+				if (this._oldKernel && this._oldKernel.name !== displayName) {
+					if (this._kernelDisplayNameToNotebookProviderIds.has(displayName)) {
+						let providerId = this._kernelDisplayNameToNotebookProviderIds.get(displayName);
+						if (providerId) {
+							if (providerId !== this._providerId) {
+								this._providerId = providerId;
+								this._onProviderIdChanged.fire(this._providerId);
```

Comment:
Does it matter if we fire onProviderIdChanged before or after we start the new session? Just curious.

Reply:
good question. I don't know. I was thinking to put startSession into handleProviderIdChanged(). But somehow not working correctly. 

Do you think it is good idea to put startSession into handleProviderIdChanged()?